src/transformers/modeling_tf_utils.py

```diff
@@ -216,77 +219,183 @@ class TFMaskedLanguageModelingLoss(TFCausalLanguageModelingLoss):
-                    "/".join(symbolic_weight.name.split("/")[2:]) for symbolic_weight in symbolic_weights
-                )
-                missing_layers.extend(list(symbolic_weights_names - saved_weight_names_set))
-                unexpected_layers.extend(list(saved_weight_names_set - symbolic_weights_names))
+                saved_weight_names_values = {}
 
-    return missing_layers, unexpected_layers
+                for weight_name in saved_weight_names:
+                    name = "/".join(weight_name.split("/")[1:])
+                    saved_weight_names_values[name] = np.asarray(g[weight_name])
+
+                for symbolic_weight in symbolic_weights:
+                    splited_layers = symbolic_weight.name.split("/")[1:]
+                    symbolic_weight_name = "/".join(splited_layers)
+
+                    if symbolic_weight_name in saved_weight_names_values:
+                        saved_weight_value = saved_weight_names_values[symbolic_weight_name]
+
+                        if K.int_shape(symbolic_weight) != saved_weight_value.shape:
+                            try:
+                                array = np.reshape(saved_weight_value, K.int_shape(symbolic_weight))
+                            except AssertionError as e:
+                                e.args += (K.int_shape(symbolic_weight), saved_weight_value.shape)
+                                raise e
+                        else:
+                            array = saved_weight_value
+
+                        weight_value_tuples.append((symbolic_weight, array))
+
+    K.batch_set_value(weight_value_tuples)
 
 
 def load_tf_weights(model, resolved_archive_file):
     """
-    Load the TF weights from a H5 file.
+    Detect missing and unexpected layers.
 
     Args:
         model (:obj:`tf.keras.models.Model`):
             The model to load the weights into.
         resolved_archive_file (:obj:`str`):
             The location of the H5 file.
+
+    Returns:
+        Two lists, one for the missing layers, and another one for the unexpected layers.
     """
+    missing_layers = []
+    unexpected_layers = []
+
     with h5py.File(resolved_archive_file, "r") as f:
-        saved_layer_names = set(hdf5_format.load_attributes_from_hdf5_group(f, "layer_names"))
-        weight_value_tuples = []
+        saved_h5_model_layers_name = set(hdf5_format.load_attributes_from_hdf5_group(f, "layer_names"))
+        model_layers_name_value = {}
 
         for layer in model.layers:
-            if layer.name in saved_layer_names:
-                g = f[layer.name]
+            name = layer.name
+            model_layers_name_value[name] = layer
+
+        model_layers_name = set(model_layers_name_value.keys())
+        renamed_saved_h5_model_layers_names = set()
+
+        for layer_name in saved_h5_model_layers_name:
+            name = layer_name.replace("nsp___", "").replace("mlm___", "")
+
+            renamed_saved_h5_model_layers_names.add(name)
```

Comment:
The more I think about this, the more I think this is unmaintainable. This is not something that should be handled directly in `modeling_tf_utils.py` imo. If this part isn't necessary for better performance, I would vote to revert that. What is the main reason you wanted to do this change?

Reply:
I can become unmaintainable, but let's see first if the other models needs changes otherwise if they don't, I don't see the problem to keep two exceptions.

Anyway revert will be easy, it will be the same code but without the two `.replace()`.

Reply:
I also agree with @Lysandre here.
1) changing weight parameters is super dangerous IMO; all kinds of problem can come up here. *E.g.* how do we deal with people training a model with new weight names of transformer v.4.0, but for some reason having to deploy the model in v3.4? We've had all kinds of issues when we added the `position_ids` as buffered weights in PyTorch with people that used different "model versions" and library versions.
2) It's just a cosmetic name change at the cost of adding having to add silent string renaming in the load weights function. I don't think it's worth it here.
3) This PR is getting huge already and will be hard to review -> can we maybe leave the name changes out of this PR and discuss it in a new one?

Reply:
1. This new way is 100% backward compatible.

2. Not only cosmetic but also for naming and better layers design.

3. This is the tyniest part of this PR.

I case it is not clear enough (my bad for this). The load_weights method will not change at all, only the occurences of `replaces` will be removed, all the rest is necessary.